### PR TITLE
Make array marshalling pin paths use array IL instructions

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Change the pinning fast paths for marshalling .NET arrays to native arrays to use array IL instructions (like how Roslyn compiles C# pinning) instead of getting the offset of the array data in the CLR object from the runtime.

This cleans up the generated IL and makes it more understandable/straightforward.

There is one behavior change: if you pass a zero-length array, such as `new int[0]`, to native the behavior has changed from passing to native a pointer to the offset into the object (which you can't write to without possibly overwriting another CLR object that happens to be next to the array in the object heap) to passing a null pointer to native. I believe this behavior change is ok since there are no safe operations possible with the current behavior and all this change would do is expose user bugs that could destabilize the runtime.

### CODE REVIEW

File: src/vm/ilmarshalers.cpp [hunk 1]
```diff
@@ -3807,9 +3803,13 @@ void ILNativeArrayMarshaler::EmitMarshalArgumentCLRToNative()
         EmitLoadManagedValue(m_pcsMarshal);
         m_pcsMarshal->EmitSTLOC(dwPinnedLocal);
         m_pcsMarshal->EmitLDLOC(dwPinnedLocal);
+        m_pcsMarshal->EmitLDLEN();
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.cpp]: I think most of this is straight forward C#, but I really like the fact that much of this IL has comments in C#.

**Comment by AaronRobinsonMSFT** [comment]
> I do not think that this behavior is ok. The Windows APIs in particular are pretty sensitive to NULL vs. non-NULL, even if the buffer is empty.

I fully agree with this concern, but at the same time would like to see determine the fallout of this change. The Win32 API is indeed finicky around buffer pointers, but most of the time the semantics are if the length is 0, then the buffer pointer itself is ignored. Most Win32 API I have used only touch the buffer pointer when when the length is > 0. Therefore I would like to try it out and if this really becomes an issue we can easily revert the change.

@jkoritzinsky Is there a way to make this change so it would be easy to revert in the case where someone hits a blocking issue?

**Comment by jkoritzinsky** [comment]
Sure! I also think I can make this change automatically revert to the old behavior if the array length is 0 with just adding to the emitted IL. It might be marginally slower (would do the offset calculation for element 0 at marshal time instead of at IL emit time) but I think in this edge case that would be fine.

If we wanted/needed to keep the calculation at IL emit time, we might as well leave the code as-is since we'd effectively have to keep all of the old code as a fallback path.

**Comment by jkotas** [comment]
I do not see the rationale behind changing the behavior here. Yes, we would design this differently if we are doing this from scratch, but that is not a good enough reason for changing it.

**Comment by jkoritzinsky** [comment]
The test failures in CI is actually unrelated to the changed IL instructions/behavior. All of the test failures are related to an incorrect assert that assumes that the `LocalDesc` for a managed type has only one component because of a difficiency in the implementation of `LocalDesc::IsValueClass`. If we change the managed type to correctly be an array, then that assert is hit.

**Comment by jkoritzinsky** [comment]
The rationale behind is PR is making it easier to understand for onboarding new people onto interop, matching the C# compiler so if we decide to make a new interop stack in C# we can make the behavior match without needing to drop to custom IL.

**Comment by jkotas** [comment]
It is fine for a new interop stack in C# to have different behaviors. But that is not a good reason for introducing a breaking changes into the existing interop stack.

**Comment by jkotas** [comment]
What about these CoreFX failures:

```
System.ArgumentException : Parameter is not valid.
+++++++++++++++++++
STACK TRACE:
at System.Drawing.StringFormat.SetTabStops(Single firstTabOffset, Single[] tabStops) at System.Drawing.Tests.StringFormatTests.SetTabStops_GetTabStops_ReturnsExpected(Single firstTabOffset, Single[] tabStops) 
```

Is it one of those cases of Windows APIs sensitive to NULL vs. non-NULL for empty buffers?

**Comment by AaronRobinsonMSFT** [comment]
@jkotas yes... Source indicates an error is thrown regardless of the `count` argument.

@jkoritzinsky This is probably going to be a mess because almost all of the GDI functions do this check, which means WinForms will be a nightmare. Based on this being so prevalent in a key sub-system, I am inclined to agree with Jan that we shouldn't invest in this work any more. Thoughts?

**Comment by jkoritzinsky** [comment]
Ok sounds good. I'll put out a PR that doesn't change behavior and close this one once that one is out.

**Pull request closed without merging**